In [1]:
import os
import pandas as pd

In [2]:
os.chdir("F:\DS tut\datasets")

In [3]:
#importing data
cars_data = pd.read_csv("Toyota.csv", index_col=0, na_values=["??", "????"])

In [4]:
cars_data.head()

Price   Age       KM FuelType    HP  MetColor  Automatic    CC  Doors  \
0  13500  23.0  46986.0   Diesel  90.0       1.0          0  2000  three   
1  13750  23.0  72937.0   Diesel  90.0       1.0          0  2000      3   
2  13950  24.0  41711.0   Diesel  90.0       NaN          0  2000      3   
3  14950  26.0  48000.0   Diesel  90.0       0.0          0  2000      3   
4  13750  30.0  38500.0   Diesel  90.0       0.0          0  2000      3   

   Weight  
0    1165  
1    1165  
2    1165  
3    1165  
4    1170

In [5]:
#copy of the original data, not related to original data
cars_data2 = cars_data.copy()

## Frequency Tables
Frequency refers to the no. of times an event or a value occurs. 
Frequency table is a table that lists items and shows the no. of times the items occur.

### Frequency table = Categories with count
### Relative frequency table = Percentage in each category i.e. 
### (freq. of an item)*100 / (Total freq.)

Total freq. = sum of all the frequencies

Percentages sum to 100%.

**pandas.crosstab():**
- To compute a simple cross-tabulation of one, two (or more) factors
- By default, computes a frequency table of the factors

Cross tabulation is usually performed on categorical data — data that can be divided into mutually exclusive groups.

An example of categorical data is the region of sales for a product. Typically, region can be divided into categories such as geographic area (North, South, Northeast, West, etc) or state (Andhra Pradesh, Rajasthan, Bihar, etc). The important thing to remember about categorical data is that a categorical data point cannot belong to more than one category.

In [6]:
'''
getting the frequencies of categories that as are available in the variable 'FuelType'
from the df 'cars_data2', 
columns -> column name that you give to the freq. column(s)
since the values that are not availale cannot be considered in
the count, we write 'dropna=True'.

1436-original data
1336-after dropping nan values
'''
pd.crosstab(index=cars_data2['FuelType'], columns='count', dropna=True)

col_0     count
FuelType       
CNG          15
Diesel      144
Petrol     1177

### Two way table

To look at the freq. of distribution of gearbox types w.r.t. diff. fuel type of the cars.

0 - Manual gear box

1 - automatic gear box

In [7]:
pd.crosstab(index=cars_data2['Automatic'], columns=cars_data2['FuelType'], dropna=True)

FuelType   CNG  Diesel  Petrol
Automatic                     
0           15     144    1104
1            0       0      73

### Two-way table-joint probability

Joint prob. is the likelihood of 2 independent events happenning at the same time.

In [8]:
pd.crosstab(index=cars_data2['Automatic'], columns=cars_data2['FuelType'], normalize=True, dropna=True)

FuelType        CNG    Diesel    Petrol
Automatic                              
0          0.011228  0.107784  0.826347
1          0.000000  0.000000  0.054641

### Two-way table-marginal probability

Marginal prob. is the prob. of the occurrence of the single event.

Setting normalize=True, we get the row sum and column sum in the column named 'All'.

(prob. CNG+Diesel+Petrol)

In [9]:
pd.crosstab(index=cars_data2['Automatic'], columns=cars_data2['FuelType'], margins=True, normalize=True, dropna=True)

FuelType        CNG    Diesel    Petrol       All
Automatic                                        
0          0.011228  0.107784  0.826347  0.945359
1          0.000000  0.000000  0.054641  0.054641
All        0.011228  0.107784  0.880988  1.000000

Look at row 1 in the above output.

0.945359 is the prob. of cars having manual gear box when the fuel type are CNG or Diesel or Petrol.

### Two-way table - conditional probability
• Conditional probability is the probability of an event ( A ), given that another event ( B ) has already occurred

• Given the type of gear box, probability of different fuel type

In [10]:
pd.crosstab(index=cars_data2['Automatic'], columns=cars_data2['FuelType'], margins=True, dropna=True, normalize='index')

FuelType        CNG    Diesel    Petrol
Automatic                              
0          0.011876  0.114014  0.874109
1          0.000000  0.000000  1.000000
All        0.011228  0.107784  0.880988

in the above output, row sum=1

In [11]:
pd.crosstab(index=cars_data2['Automatic'], columns=cars_data2['FuelType'], margins=True, dropna=True, normalize='columns')

FuelType   CNG  Diesel    Petrol       All
Automatic                                 
0          1.0     1.0  0.937978  0.945359
1          0.0     0.0  0.062022  0.054641

in the above output, column sum = 1 

## Correlation

checking relation betn 2 numerical variables.

(variables might not always be numerical, but in this case they are numerical)

- Visual representation of corrrelation: Scatter plots

range of correlation: -1 to 1

-1 or closer -> -ve correlation

+1 or closer -> high +ve correlation

0 -> no correlation

DataFrame.corr(self, method='pearson’)

• To compute pairwise correlation of columns excluding NA/null
values

• Excluding the categorical variables to find the Pearson’s
correlation

In [12]:
numerical_data = cars_data2.select_dtypes(exclude=[object])

In [13]:
#no. of variables available under numerical_data
print(numerical_data.shape)

(1436, 8)


Correlation between numerical variables

Command: 

DataFrame.corr(self, method='pearson')

In [14]:
corr_matrix = numerical_data.corr()